File: usage/crunch.py
#!/usr/bin/env python3
"""Usage crunch — pings CSV + roster → the uid-keyed DATA blob for usage/index.html.

Reads the *published-to-web* CSV of the analytics sheet's pings tab (the sheet itself
stays private; unpublishing that tab breaks this) and the public roster gviz. Names
never land in the output: people are keyed by their ping uid (first 8 hex of
SHA-256(name)) and the page joins names back at runtime from the roster. Run:

    python3 usage/crunch.py out.json
"""
import csv, hashlib, io, json, re, sys, urllib.request
from datetime import datetime, timedelta

CSV = ("https://docs.google.com/spreadsheets/d/e/2PACX-1vRXFSKEV0IjMxBH-hSWAvirawzMmSBm4rEtj0W7vTM8OsH3yDZdK_Z4bjpgDXY6PBw8DvdPoM5E3lUo"
       "/pub?gid=0&single=true&output=csv")
SID, GID = "1j__RMUvFWQlX9UuT-Uxkw7BkqWHCQkbR_hKsTyNwiyo", "800090339"   # keep in sync with roster-data.js
LAUNCH = datetime(2026, 7, 6, 19, 30)   # shared over dinner; earlier rows = testing
END = datetime(2026, 7, 13, 0, 0)        # festival's over: the frozen dashboard counts only [LAUNCH, END);
                                         # later opens go to the separate post-fest block, never diluting it
JASON = "70f71792"                       # uid("Jason Sundram") — shown, not ranked
QUEUED = timedelta(seconds=90)           # received this far after opened = was offline

uid = lambda name: hashlib.sha256(name.encode()).hexdigest()[:8]
fetch = lambda url: urllib.request.urlopen(url, timeout=30).read().decode()

def roster_uids():
    """uid set of everyone on the roster (gviz header quirk: labels or row 0)."""
    raw = fetch(f"https://docs.google.com/spreadsheets/d/{SID}/gviz/tq?gid={GID}&tqx=out:json")
    t = json.loads(raw[raw.index("{"): raw.rindex("}") + 1])["table"]
    rows = [[(c or {}).get("v") or "" for c in r["c"]] for r in t["rows"]]
    labels = [(c.get("label") or "").lower() for c in t["cols"]]
    if not any("name" in l for l in labels):
        labels, rows = [str(x).lower() for x in rows[0]], rows[1:]
    ni = next(i for i, l in enumerate(labels) if "name" in l)
    names = [str(r[ni]).strip() for r in rows if str(r[ni]).strip().lower() not in ("", "name")]
    return {uid(n) for n in names}

def T(s):
    m = re.match(r"(\d+)/(\d+)/(\d+) (\d+):(\d+):(\d+)", s or "")
    return datetime(int(m[3]), int(m[1]), int(m[2]), int(m[4]), int(m[5]), int(m[6])) if m else None

def kudos_crunch(krows):
    """The 👏 applause events (empty page, action=kudos): who was cheered, for which piece.
    Same interrupted-flush dedup + launch filter as opens; recipients/senders stay uid-keyed
    (joined to names at runtime), composer labels are non-PII plaintext."""
    seen, dd = set(), []
    for x in krows:                                   # dedup on (opened, sender, recipient, composer)
        if x not in seen: seen.add(x); dd.append(x)
    dd = [x for x in dd if x[0] and LAUNCH <= x[0] < END]
    to_ct, comp_ct, senders = {}, {}, set()
    for opened, who, to, label in dd:
        if who: senders.add(who)
        if to: to_ct[to] = to_ct.get(to, 0) + 1
        if label: comp_ct[label] = comp_ct.get(label, 0) + 1
    return {
        "total": len(dd), "senders": len(senders), "recipients": len(to_ct),
        "byComposer": sorted(({"label": l, "n": n} for l, n in comp_ct.items()),
                             key=lambda c: (-c["n"], c["label"])),
        "toList": sorted(({"uid": u, "n": n} for u, n in to_ct.items()),
                         key=lambda t: (-t["n"], t["uid"])),
    }

def post_crunch(post_rows):
    """Post-festival trickle — deliberately OUT of the frozen dashboard, surfaced on its own:
    opens, unique identified users, anon opens, and a by-day count. Enough to answer 'still used?'
    without letting quiet post-fest days extend the adoption/rhythm charts."""
    users, by_day, anon = set(), {}, 0
    for opened, rec, page, who in post_rows:
        d = f"{opened.month}/{opened.day}"
        by_day[d] = by_day.get(d, 0) + 1
        if who: users.add(who)
        else: anon += 1
    return {"opens": len(post_rows), "users": len(users), "anon": anon, "byDay": by_day}

def crunch(csv_text=None, roster=None):
    if roster is None: roster = roster_uids()          # inject both to test offline (also the JS-parity oracle)
    if csv_text is None: csv_text = fetch(CSV)
    rows, krows = [], []
    for r in csv.DictReader(io.StringIO(csv_text)):
        rec, opened = T(r.get("received")), T(r.get("opened")) or T(r.get("received"))
        page, who = (r.get("page") or "").strip(), (r.get("who") or "").strip()
        if rec and page and page != "page":
            rows.append((opened, rec, page, who))
        elif rec and (r.get("action") or "").strip() == "kudos":     # a feature-use event, empty page
            krows.append((opened, who, (r.get("to") or "").strip(), (r.get("label") or "").strip()))
    rows.sort()
    raw = len(rows)
    seen, dd = set(), []
    for x in rows:                                   # a flush interrupted mid-save re-sends
        k = (x[0], x[2], x[3])
        if k not in seen: seen.add(k); dd.append(x)
    dupes = raw - len(dd)                             # duplicate deliveries removed (before any date split)
    pre = sum(1 for x in dd if x[0] < LAUNCH)
    post_rows = [x for x in dd if x[0] >= END]
    dd = [x for x in dd if LAUNCH <= x[0] < END]

    day = lambda d: f"{d.month}/{d.day}"
    launch_day = day(LAUNCH)                 # the launch evening's burst skews the hour-of-day rhythm
    users, pages, by_day, anon_day = {}, {}, {}, {}
    by_hour = [0] * 24
    buckets, first = {}, {}
    anon = queued = 0
    for opened, rec, page, who in dd:
        d, h = day(opened), opened.hour
        by_day[d] = by_day.get(d, 0) + 1
        if d != launch_day: by_hour[h] += 1     # rhythm of day leaves out launch night's 19:30 spike
        buckets[(d, h)] = buckets.get((d, h), 0) + 1
        pg = pages.setdefault(page, {"opens": 0, "users": set()})
        pg["opens"] += 1
        if rec - opened > QUEUED: queued += 1
        if not who:
            anon += 1; anon_day[d] = anon_day.get(d, 0) + 1; continue
        pg["users"].add(who)
        u = users.setdefault(who, {"uid": who, "opens": 0, "pages": {}, "days": set()})
        u["opens"] += 1
        u["pages"][page] = u["pages"].get(page, 0) + 1
        u["days"].add(d)
        if who != JASON: first.setdefault(who, opened)

    series, adopt = [], []
    t, end = LAUNCH.replace(minute=0), max(x[0] for x in dd)
    firsts = sorted(first.values())
    while t <= end:
        series.append({"t": f"{day(t)} {t.hour:02d}", "n": buckets.get((day(t), t.hour), 0)})
        adopt.append({"t": series[-1]["t"], "n": sum(1 for f in firsts if f < t + timedelta(hours=1))})
        t += timedelta(hours=1)

    ulist = sorted(users.values(), key=lambda u: -u["opens"])
    return {
        "generated": datetime.now().strftime("%-m/%-d %H:%M"),
        "raw": raw, "deduped": dupes, "pre": pre, "total": len(dd),
        "anon": anon, "queued": queued,
        "rosterSize": len(roster),
        "identified": len(ulist),
        "unknown": sum(1 for u in ulist if u["uid"] not in roster),
        "users": [{"uid": u["uid"], "opens": u["opens"], "pages": u["pages"],
                   "days": len(u["days"]), "known": u["uid"] in roster} for u in ulist],
        "pages": sorted([{"page": p, "opens": v["opens"], "users": len(v["users"])}
                         for p, v in pages.items()], key=lambda p: -p["opens"]),
        "byDay": by_day, "anonByDay": anon_day, "byHour": by_hour,
        "series": series, "adopt": adopt,
        "kudos": kudos_crunch(krows),
        "post": post_crunch(post_rows),
    }

if __name__ == "__main__":
    try:
        data = crunch()
    except OSError as e:
        sys.exit(f"fetch failed (offline? unpublished?): {e}")
    out = json.dumps(data, separators=(",", ":"))
    (open(sys.argv[1], "w") if len(sys.argv) > 1 else sys.stdout).write(out)

File: usage/crunch.js
// Usage crunch, in the browser — a 1:1 port of usage/crunch.py so the /usage/ page
// refreshes itself live (open = refresh) instead of waiting on the nightly laptop job.
// Same shape out as the old baked DATA blob; scripts/usage-test.js pins it to the
// Python oracle's golden so the two can't drift.
//
// Input is the analytics sheet's pings tab (published-to-web CSV, CORS-clean) + the
// set of roster uids (SHA-256(name)[:8]) the page already derives to join names. Names
// never touch this file — everyone is keyed by uid, exactly as in the CSV and the blob.
(function () {
  const JASON = "70f71792";                                  // uid("Jason Sundram") — shown, not ranked
  const LAUNCH = Date.UTC(2026, 6, 6, 19, 30);               // shared over dinner; earlier rows = testing
  const END = Date.UTC(2026, 6, 13, 0, 0);                   // festival's over: frozen dashboard = [LAUNCH, END); later opens → the post-fest block
  const SERIES0 = Date.UTC(2026, 6, 6, 19, 0);               // LAUNCH.replace(minute=0)
  const QUEUED = 90e3;                                        // received this far after opened = was offline
  const HOUR = 36e5;

  // wall-clock, tz-agnostic: parse "M/D/YYYY H:M:S" into a UTC instant so bucketing by
  // day/hour and hourly stepping never touch local time or DST — matching Python's naive
  // datetimes (the viewer could be in any timezone; the sheet's clock is Vienna's).
  const T = s => {
    const m = /^(\d+)\/(\d+)\/(\d+) (\d+):(\d+):(\d+)/.exec(s || "");
    return m ? Date.UTC(+m[3], +m[1] - 1, +m[2], +m[4], +m[5], +m[6]) : null;
  };
  const day = ms => { const d = new Date(ms); return `${d.getUTCMonth() + 1}/${d.getUTCDate()}`; };
  const hour = ms => new Date(ms).getUTCHours();

  // minimal RFC-4180 CSV: quoted fields, "" escapes, commas/newlines inside quotes.
  function parseCSV(text) {
    const rows = [[]]; let f = "", q = false;
    text = text.replace(/\r\n?/g, "\n");
    for (let i = 0; i < text.length; i++) {
      const c = text[i];
      if (q) {
        if (c === '"') { if (text[i + 1] === '"') { f += '"'; i++; } else q = false; }
        else f += c;
      } else if (c === '"') q = true;
      else if (c === ",") { rows[rows.length - 1].push(f); f = ""; }
      else if (c === "\n") { rows[rows.length - 1].push(f); f = ""; rows.push([]); }
      else f += c;
    }
    rows[rows.length - 1].push(f);
    if (rows.length && rows[rows.length - 1].length === 1 && rows[rows.length - 1][0] === "") rows.pop();
    if (!rows.length) return [];
    const head = rows[0].map(h => h.trim());
    return rows.slice(1).map(r => { const o = {}; head.forEach((h, i) => o[h] = r[i]); return o; });
  }

  // the 👏 applause events (empty page, action=kudos): who was cheered, for which piece.
  // same interrupted-flush dedup + launch filter as opens; recipients/senders stay uid-keyed.
  function kudosCrunch(krows) {
    const seen = new Set(), dd = [];                          // dedup on (opened, sender, recipient, composer)
    for (const x of krows) { const k = `${x.opened}|${x.who}|${x.to}|${x.label}`;
      if (!seen.has(k)) { seen.add(k); dd.push(x); } }
    const kept = dd.filter(x => x.opened !== null && x.opened >= LAUNCH && x.opened < END);
    const toCt = new Map(), compCt = new Map(), senders = new Set();
    for (const { who, to, label } of kept) {
      if (who) senders.add(who);
      if (to) toCt.set(to, (toCt.get(to) || 0) + 1);
      if (label) compCt.set(label, (compCt.get(label) || 0) + 1);
    }
    const scmp = (a, b) => a < b ? -1 : a > b ? 1 : 0;
    return {
      total: kept.length, senders: senders.size, recipients: toCt.size,
      byComposer: [...compCt.entries()].map(([label, n]) => ({ label, n }))
        .sort((a, b) => b.n - a.n || scmp(a.label, b.label)),
      toList: [...toCt.entries()].map(([uid, n]) => ({ uid, n }))
        .sort((a, b) => b.n - a.n || scmp(a.uid, b.uid)),
    };
  }

  // post-festival trickle — deliberately OUT of the frozen dashboard, surfaced on its own:
  // opens, unique identified users, anon opens, and a by-day count. Answers "still used?".
  function postCrunch(postRows) {
    const users = new Set(), byDay = {}; let anon = 0;
    for (const { opened, who } of postRows) {
      const d = day(opened);
      byDay[d] = (byDay[d] || 0) + 1;
      if (who) users.add(who); else anon++;
    }
    return { opens: postRows.length, users: users.size, anon, byDay };
  }

  function crunch(csvText, rosterUids, now = new Date()) {
    const roster = rosterUids instanceof Set ? rosterUids : new Set(rosterUids || []);
    let rows = [], krows = [];
    for (const r of parseCSV(csvText)) {
      const rec = T(r.received), opened = T(r.opened) || T(r.received);
      const page = (r.page || "").trim(), who = (r.who || "").trim();
      if (rec !== null && page && page !== "page") rows.push({ opened, rec, page, who });
      else if (rec !== null && (r.action || "").trim() === "kudos")    // a feature-use event, empty page
        krows.push({ opened, who, to: (r.to || "").trim(), label: (r.label || "").trim() });
    }
    // sort by (opened, rec, page, who) — same tuple order Python sorts, so dedup keeps
    // the same first occurrence and every insertion-ordered dict below matches
    const cmp = (a, b) => a - b || 0;
    const scmp = (a, b) => a < b ? -1 : a > b ? 1 : 0;
    rows.sort((a, b) => cmp(a.opened, b.opened) || cmp(a.rec, b.rec) || scmp(a.page, b.page) || scmp(a.who, b.who));
    const raw = rows.length;

    const seen = new Set(), dedup = [];                        // a flush interrupted mid-save re-sends
    for (const x of rows) {
      const k = `${x.opened} ${x.page} ${x.who}`;
      if (!seen.has(k)) { seen.add(k); dedup.push(x); }
    }
    const dupes = raw - dedup.length;                         // duplicate deliveries removed (before any date split)
    const pre = dedup.filter(x => x.opened < LAUNCH).length;
    const postRows = dedup.filter(x => x.opened >= END);
    const dd = dedup.filter(x => x.opened >= LAUNCH && x.opened < END);

    const launchDay = day(LAUNCH);                             // launch night skews the hour-of-day rhythm
    const users = new Map(), pages = new Map(), byDay = {}, anonDay = {};
    const byHour = new Array(24).fill(0);
    const buckets = new Map(), first = new Map();
    let anon = 0, queued = 0;
    for (const { opened, rec, page, who } of dd) {
      const d = day(opened), h = hour(opened);
      byDay[d] = (byDay[d] || 0) + 1;
      if (d !== launchDay) byHour[h]++;                        // rhythm of day leaves out launch night's spike
      const bk = `${d} ${h}`; buckets.set(bk, (buckets.get(bk) || 0) + 1);
      let pg = pages.get(page); if (!pg) pages.set(page, pg = { opens: 0, users: new Set() });
      pg.opens++;
      if (rec - opened > QUEUED) queued++;
      if (!who) { anon++; anonDay[d] = (anonDay[d] || 0) + 1; continue; }
      pg.users.add(who);
      let u = users.get(who); if (!u) users.set(who, u = { uid: who, opens: 0, pages: {}, days: new Set() });
      u.opens++;
      u.pages[page] = (u.pages[page] || 0) + 1;
      u.days.add(d);
      if (who !== JASON && !first.has(who)) first.set(who, opened);
    }

    const series = [], adopt = [];
    const end = Math.max(...dd.map(x => x.opened));
    const firsts = [...first.values()].sort((a, b) => a - b);
    for (let t = SERIES0; t <= end; t += HOUR) {
      const label = `${day(t)} ${String(hour(t)).padStart(2, "0")}`;
      series.push({ t: label, n: buckets.get(`${day(t)} ${hour(t)}`) || 0 });
      adopt.push({ t: label, n: firsts.filter(f => f < t + HOUR).length });
    }

    const ulist = [...users.values()].sort((a, b) => b.opens - a.opens);
    const pad2 = n => String(n).padStart(2, "0");
    return {
      generated: `${now.getMonth() + 1}/${now.getDate()} ${pad2(now.getHours())}:${pad2(now.getMinutes())}`,
      raw, deduped: dupes, pre, total: dd.length,
      anon, queued,
      rosterSize: roster.size,
      identified: ulist.length,
      unknown: ulist.filter(u => !roster.has(u.uid)).length,
      users: ulist.map(u => ({ uid: u.uid, opens: u.opens, pages: u.pages, days: u.days.size, known: roster.has(u.uid) })),
      pages: [...pages.entries()].map(([p, v]) => ({ page: p, opens: v.opens, users: v.users.size }))
        .sort((a, b) => b.opens - a.opens),
      byDay, anonByDay: anonDay, byHour,
      series, adopt,
      kudos: kudosCrunch(krows),
      post: postCrunch(postRows),
    };
  }

  const api = { crunch, parseCSV };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  else window.UsageCrunch = api;
})();
